pgb_utils/tutorials/PGB_Tutorial.ipynb

```diff
@@ -0,0 +1,2756 @@
+            "tags": []
+          },
+          "execution_count": 12
+        }
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "zNuQD35Zq7yJ"
+      },
+      "source": [
+        "pgb.bigquery.get_table_info('candidates')\n",
+        "\n",
+        "# Colab Hint: Right-click this cell and select \"Copy to scratch cell\"\n",
+        "#             so you can use this as a reference later."
+      ],
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "cdy9BeSKT60F"
+      },
+      "source": [
+        "---"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "AbTyVXUko6-d"
+      },
+      "source": [
+        "### Query lightcurves and other history"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "-7VY4RP6748V"
+      },
+      "source": [
+        "# Choose the history data you want returned\n",
+        "columns = ['jd', 'fid', 'magpsf', 'sigmapsf']\n",
+        "# 'objectId' and 'candid' will be included automatically\n",
+        "# options are in the 'candidates' table\n",
+        "# pgb.bigquery.get_table_info('candidates')\n",
+        "\n",
+        "# Choose specific objects (optional, but reduces load time)\n",
+        "objectIds = ['ZTF18aczuwfe', 'ZTF18aczvqcr', 'ZTF20acqgklx', 'ZTF18acexdlh']"
+      ],
+      "execution_count": 15,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "V4JfqzMTXIpk"
+      },
+      "source": [
+        "To retrieve lightcurves and other history, we must query for the objects' \"candidate\" observations and aggregate the results by `objectId`.\n",
+        "- `pgb.bigquery.query_objects()` is a convenience wrapper for this.\n",
+        "It's options are demonstrated below. \n",
+        "- You can also use `pgb.bigquery.object_history_sql_statement()` to get the required SQL statement and make the query yourself. See the \"Using `google.cloud.bigquery` directly\" section."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "YtJ2C1gS9vdC"
+      },
+      "source": [
+        "# Option 1: Get a single DataFrame of all results\n",
+        "\n",
+        "lcsdf = pgb.bigquery.query_objects(columns, objectIds=objectIds)\n",
+        "# This will execute a dry run and tell you how much data will be processed.\n",
+        "# You will be asked to confirm before proceeding.\n",
+        "# In the future we'll skip this using\n",
+        "dry_run = False\n",
+        "\n",
+        "lcsdf.sample(10)\n",
+        "# cleaned of duplicates"
+      ],
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "kEBGXlyM_P33"
+      },
+      "source": [
+        "# Option 2: Get a generator that yields a DataFrame for each objectId\n",
```

Comment:
The answer seems obvious to me, so I'm guessing there are implied subtleties that I don't understand. My previous generator experience is only as a naive user.

My answer:
So that people can iterate through dataframes of individual objects nicely, and not have to load all of the query results at once, which could take up a significant amount of memory.

Reply:
1. How much memory does it take up?
2. I'll bet that the query results are all being loaded into memory anyway.

I think there are reasons to use generators, but as a simple rule of thumb I think you need to be talking about at least hundreds of megabytes before it really makes sense.

Reply:
> 1. How much memory does it take up?
> I think there are reasons to use generators, but as a simple rule of thumb I think you need to be talking about at least hundreds of megabytes before it really makes sense.

I don't know how much memory our current tables would take up to load in this way (to a dataframe), but I was thinking ahead to LSST-scale data. I was hoping to check how much _time_ it would take to load the entire candidates table and then leave a comment in the notebook (if I could find the time before the workshop). I can check the memory too. In BigQuery, our alerts table is currently 315 GB, so I'm guessing the candidates table will be about 30 GB when I create it. But I don't know how that maps to the amount of memory when loaded locally.

> 2. I'll bet that the query results are all being loaded into memory anyway.

I was convinced that they were not, and that was another reason I wrote the generator option. But in looking again just now, I'm less sure. I can see how they might be. I will try to just check how much memory the `query_job` object takes up.

Reply:
@wmwv, just FYI: 
I am again convinced that Google SDK returns a `query_job` object that does not load all the query results into memory. Here are some checks and their results, for reference.

```python
columns = ['magpsf','sigmapsf','fid']

# Check the raw query_job object
print('checking query_job')
format='query_job'

limit=10
query_job = pgb.bigquery.query_objects(columns, limit=limit, format=format, dry_run=False)
print(sys.getsizeof(query_job))

limit=5000
query_job = pgb.bigquery.query_objects(columns, limit=limit, format=format, dry_run=False)
print(sys.getsizeof(query_job))

# Check the dataframes
print('checking pandas dataframes')
format='pandas'
iterator=False

limit=10
df = pgb.bigquery.query_objects(columns, limit=limit, format=format, iterator=iterator, dry_run=False)
print(sys.getsizeof(df))

limit=5000
df = pgb.bigquery.query_objects(columns, limit=limit, format=format, iterator=iterator, dry_run=False)
print(sys.getsizeof(df))

# Check the dataframe generators
print('checking pandas df generators')
format='pandas'
iterator=True

limit=10
df_gntr = pgb.bigquery.query_objects(columns, limit=limit, format=format, iterator=iterator, dry_run=False)
print(sys.getsizeof(df_gntr))

limit=5000
df_gntr = pgb.bigquery.query_objects(columns, limit=limit, format=format, iterator=iterator, dry_run=False)
print(sys.getsizeof(df_gntr))
```

Output:
```python
checking query_job
64
64
checking pandas dataframes
23679
1804712
checking pandas df generators
128
128
```

Reply:
Cool thanks.  If you have time, add a bit of a note explaining that you can grab them all as a list, or you can step through them using a generator.  And that further using a generator would be important for larger data scales and thus, here's how to do it.